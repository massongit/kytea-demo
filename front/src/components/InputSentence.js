import React from "react"
import fetch from "node-fetch"
import PropTypes from "prop-types"
import {Button, Form, Input, InputGroup, InputGroupAddon} from "reactstrap"
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome"
import {FormattedMessage, intlShape} from "react-intl"
import {loading, showSentence} from "../actions"
import {faSpinner} from "@fortawesome/free-solid-svg-icons"

/**
 * 入力部
 */
class InputSentence extends React.Component {
    static propTypes = {
        dispatch: PropTypes.func.isRequired,
        sentence: PropTypes.string.isRequired,
        loading: PropTypes.bool.isRequired,
        intl: intlShape.isRequired
    }

    constructor(props) {
        super(props)
        this.onSubmit = this.onSubmit.bind(this)
    }

    /**
     * KyTeaのAPIを呼び出す
     * @param value {string} 入力文
     * @returns {Promise<*>} 解析結果
     */
    static async callKyTea(value) {
        const res = await fetch(
            "kytea",
            {
                method: "POST",
                body: value
            }
        )

        if (res.ok) {
            return await res.json()
        } else {
            throw new Error(res.statusText)
        }
    }

    /**
     * onSubmitイベント
     * @param ev イベント
     */
    async onSubmit(ev) {
        // サーバーへのSubmitが行われないようにする
        ev.preventDefault()

        this.input.value = this.input.value.trim()

        // 以前の入力文とは異なる文章が入力されたとき
        if (0 < this.input.value.length && this.input.value.replace(/\s/g, "") !== this.props.sentence) {
            // ローディングアイコンを表示する
            this.props.dispatch(loading({
                loading: true
            }))

            // KyTeaによる解析を行い、KyTeaによる解析結果の表示Actionをdispatch
            try {
                this.props.dispatch(showSentence({
                    sentence: this.input.value,
                    words: await InputSentence.callKyTea(this.input.value)
                }))
            } catch (er) {
                alert(this.props.intl.formatMessage(
                    {
                        id: "errorMessage.predict"
                    },
                    {
                        message: er.message
                    }
                ))
            }

            // ローディングアイコンを消す
            this.props.dispatch(loading({
                loading: false
            }))
        }
    }

    /**
     * ローディングアイコンを描画する
     * @returns {*} ローディングアイコン
     */
    renderSpinner() {
        if (this.props.loading) {
            return (
                <FontAwesomeIcon
                    icon={faSpinner}
                    spin
                />
            )
        }
    }

    render() {
        return (
            <Form
                onSubmit={this.onSubmit}
            >
                <InputGroup className="mb-3">
                    <Input
                        innerRef={n => {
                            this.input = n
                        }}
                    />
                    <InputGroupAddon addonType="append">
                        <Button
                            type="submit"
                            color="primary"
                            disabled={this.props.loading}
                        >
                            {this.renderSpinner()}
                            <FormattedMessage id="predict"/>
                        </Button>
                    </InputGroupAddon>
                </InputGroup>
            </Form>
        )
    }
}

export default InputSentence
