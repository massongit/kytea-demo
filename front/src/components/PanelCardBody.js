import React from "react"
import PropTypes from "prop-types"
import {CardBody} from "reactstrap"
import {intlShape} from "react-intl"

/**
 * CardのBody
 */
class PanelCardBody extends React.Component {
    static propTypes = {
        messageId: PropTypes.string.isRequired,
        children: PropTypes.node,
        intl: intlShape.isRequired
    }

    /**
     * 説明部を描画する
     * @returns {*} 説明部
     */
    renderDescription() {
        if (this.props.messageId) {
            const description = this.props.intl.formatMessage({
                id: this.props.messageId
            })

            if (this.props.children) {
                return (
                    <p>
                        {description}
                    </p>
                )
            } else {
                return description
            }
        }
    }

    render() {
        return (
            <CardBody>
                {this.renderDescription()}
                {this.props.children}
            </CardBody>
        )
    }
}

export default PanelCardBody
