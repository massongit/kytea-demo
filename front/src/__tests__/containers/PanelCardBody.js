import React from "react"
import PanelCardBody from "../../containers/PanelCardBody"
import InputSentence from "../../containers/InputSentence"
import rootReducer from "../../reducers"
import {CardBody} from "reactstrap"
import {loadTranslation, mountWithIntl, shallowWithIntl} from "enzyme-react-intl"
import {createStore} from "redux"

loadTranslation("./src/translations/ja.json")

describe("containers/PanelCardBody", () => {
    it("Componentが正しく配置されている", () => {
        const panelBodyComponent = shallowWithIntl(
            <PanelCardBody messageId="description.input">
                <InputSentence/>
            </PanelCardBody>
        ).dive()
        expect(panelBodyComponent).toMatchSnapshot()
    })

    it("CardBodyの直下に子要素が入っている", () => {
        const panelBodyComponent = mountWithIntl(
            <PanelCardBody messageId="description.input">
                <InputSentence
                    store={createStore(rootReducer)}
                />
            </PanelCardBody>
        )
        expect(panelBodyComponent.find(CardBody).children().contains(InputSentence)).toBeTruthy()
    })
})
