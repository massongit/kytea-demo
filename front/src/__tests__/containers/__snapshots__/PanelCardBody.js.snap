// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/PanelCardBody Componentが正しく配置されている 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PanelCardBody
    intl={
      Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatHTMLMessage": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatPlural": [Function],
        "formatRelative": [Function],
        "formatTime": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralFormat": [Function],
          "getRelativeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {
          "abstract.kyTea": "京都テキスト解析ツールキット",
          "candidate": "候補{no}",
          "description.input": "日本語の文章を入力し、予測ボタンをクリックしてください。",
          "description.kyTea": "日本語の文章中に存在する各単語の品詞や読みを予測するシステムです。",
          "description.result": "単語を選択すると、その単語の品詞や読みが表示されます。",
          "errorMessage.default": "エラーが発生しました: {message}",
          "errorMessage.predict": "予測中にエラーが発生しました: {message}",
          "header.candidate": "候補",
          "header.margin": "SVM平面からのマージン (読み)",
          "header.pos": "品詞",
          "header.pronunciation": "読み",
          "header.word": "単語",
          "predict": "予測",
          "title.description": "概要",
          "title.input": "入力",
          "title.result": "結果",
          "unknown": "(予測不能)",
        },
        "now": [Function],
        "onError": [Function],
        "textComponent": "span",
        "timeZone": null,
      }
    }
    messageId="description.input"
  >
    <InjectIntl(Connect(InputSentence)) />
  </PanelCardBody>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <CardText
          tag="p"
        >
          日本語の文章を入力し、予測ボタンをクリックしてください。
        </CardText>,
        <InjectIntl(Connect(InputSentence)) />,
      ],
      "tag": "div",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "日本語の文章を入力し、予測ボタンをクリックしてください。",
          "tag": "p",
        },
        "ref": null,
        "rendered": "日本語の文章を入力し、予測ボタンをクリックしてください。",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <CardText
            tag="p"
          >
            日本語の文章を入力し、予測ボタンをクリックしてください。
          </CardText>,
          <InjectIntl(Connect(InputSentence)) />,
        ],
        "tag": "div",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "日本語の文章を入力し、予測ボタンをクリックしてください。",
            "tag": "p",
          },
          "ref": null,
          "rendered": "日本語の文章を入力し、予測ボタンをクリックしてください。",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "intl": Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatHTMLMessage": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatPlural": [Function],
        "formatRelative": [Function],
        "formatTime": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralFormat": [Function],
          "getRelativeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {
          "abstract.kyTea": "京都テキスト解析ツールキット",
          "candidate": "候補{no}",
          "description.input": "日本語の文章を入力し、予測ボタンをクリックしてください。",
          "description.kyTea": "日本語の文章中に存在する各単語の品詞や読みを予測するシステムです。",
          "description.result": "単語を選択すると、その単語の品詞や読みが表示されます。",
          "errorMessage.default": "エラーが発生しました: {message}",
          "errorMessage.predict": "予測中にエラーが発生しました: {message}",
          "header.candidate": "候補",
          "header.margin": "SVM平面からのマージン (読み)",
          "header.pos": "品詞",
          "header.pronunciation": "読み",
          "header.word": "単語",
          "predict": "予測",
          "title.description": "概要",
          "title.input": "入力",
          "title.result": "結果",
          "unknown": "(予測不能)",
        },
        "now": [Function],
        "onError": [Function],
        "textComponent": "span",
        "timeZone": null,
      },
    },
  },
}
`;
